### ISSUE TITLE
MO 'override' folder concept

### ISSUE THREAD

**Issue by SilverIce** [TanninOne/modorganizer#457]
The idea of the override folder is to keep mod's files 'immutable'?
I'm experiencing an issue when a program overwrites mod's files and it's really hard to find the modified files, because the files aren't in the override folder.

Real world example: a file 'head.nif' may belong to a modA, modB, modC, _plenty of other mods_. A program creates a new head.nif file. The only case when it's easy to find the file, is the case when you know which mods may contain it.

> The question turned into feature request, into something like `Would be cool to have an option to protect mod's files, that would also help isolate profiles from each other`. See below

**Comment by TanninOne** [comment]
No, the override folder is **not** intended to keep mods immutable! It's there to provide a place to store newly created files that can't automatically be assigned to an existing mod.

In your real world example, if there doesn't exist a head.nif already then it *will* be created in override because MO has no way of knowing which mod it should belong to.
If there already is a head.nif then you're replacing an existing file and MO will do just that: replace the existing file. And it's clear which head.nif gets replaced: the one with the highest priority because that's the one you read and modified.

**Comment by SilverIce** [comment]
Thanks for reply. It's clear when you know which mod includes the file. Would be cool to have an option to protect mod's files, that would also help isolate profiles from each other

Have happy New Year!

**Comment by TanninOne** [comment]
It would be cool, yes, but almost impossible to implement.Most applications, when they overwrite a file they first create a new temporary file with the new content. When this was successful they delete the file then rename the tmp file to the real file name.
MOs vfs as a file-system based tool sees this as three independent operations:
* create
* delete
* rename

To keep the file unmodified, MO would have to know in step 2 that step 3 is coming and then prevent the deletion from happening.
And even if I could write code that predicts the future, the application might check if the file is actually gone between steps 2 and 3.

As an alternative I could - instead of deleting the file - rename it (i.e. to <oldname>.deleted) but this would not leave the mod immutable and for everyone who prefers "mutable" mods it would require those files to be cleaned up manually.

**Comment by ThosRTanner** [comment]
If the mod is marked read-only, wouldn't you prevent the the delete in the first place?

**Comment by TanninOne** [comment]
What do you mean "read-only"? MO doesn't provide a read-only mechanism. Or do you mean making the directory read-only through explorer? In this case it is the OS that prevents the write with all the default error handling and however the Application deals with it. MOs behaviour doesn't change, the modification would be prevented, not redirected to Overwrite.

**Comment by ThosRTanner** [comment]
I didn't say it was. but the request was "Would be cool to have an option to protect mod's files, that would also help isolate profiles from each other", which is pretty much to provide a read-only system (in parts)

So if you in the vfs (I don't know how difficult this would be, admittedly, as I haven't read the code) determine something is from a protected mod, then the 2nd step in the 3 above - deleting a file - would be rejected whether or not step 3 is coming

**Comment by TanninOne** [comment]
The application (or whatever framework it is using) will not proceed with step 3 if step 2 failed, so the vfs would have to make the application believe the file is deleted (i.e. by removing it from the vfs without deleting the physical file) so it proceeds with the move/rename in step 3.
Humm, this should actually be very doable. I'll keep this open as a feature request.

**Comment by SilverIce** [comment]
Sorry for causing all this. I do realize, that it may cause backward compatibility issues and most of users may not need it at all

But, thank you for considering this as a feature ;)

**Comment by AnyOldName3** [comment]
This looks to me like it should work: instead of renaming the file in its original mod, write a new file to the overwrite mod that says to hide the file. This would allow the old mod to be immutable and the new mod (and any executables launched through MO) to see the file as gone.

I may have explained this poorly, so I'll include a pseudocode example for clarity:

```
void delete(relativeFilePath)
{
    writefile(OVERWRITEDIR + relativeFilePath + ".ISDELETED", null);
}

File attemptRead(relativeFilePath)
{
    for(modFolder : modFoldersAndOverwriteInPriorityOrder)
    {
        if(fileExists(modFolder + relativeFilePath))
            return readFile(modFolder + relativeFilePath);
        elseif(fileExists(modFolder + relativeFilePath + ".ISDELETED"))
            return null;
    }
    return null;
}

void write(relativeFilePath, contents)
{
    if(fileExists(OVERWRITEDIR + relativeFilePath + ".ISDELETED"))
        deleteFile(OVERWRITEDIR + relativeFilePath + ".ISDELETED");
    writeFile(OVERWRITEDIR + relativeFilePath, contents);
}
````

Of course, I'm not an expert on the guts of the VFS, but this doesn't seem overly complex and I *think* it'd solve every problem mentioned here with no compromises.

Of course, I've not given any pseudocode for the function which checks for .ISDELETED files when you add files to an existing mod.

**Comment by grasmann** [comment]
@AnyOldName3 
Renaming the files shouldn't even be necessary.
As far as I understand the main topic is that newly created files shouldn't override a mod file but instead appear in the override folder.

So if this particular option is activated / it becomes default functionality.
- Requests to delete files are always canceled and instead the files are removed from the vfs
- Newly created files are always moved to override mod regardless if they already exist in another mod or not, although within override files should be overwritten if they exist

To be honest it makes sense that this is the default.

**Comment by TanninOne** [comment]
Unfortunately it's not that easy grasmann. MO can't always know if a file is going to be written. When the application is copying or moving a file it's clear that the target file will be created fresh or replaced -> easy.
But what if the file is opened for reading/writing? The target file might not exist, then it's a new file and all is easy, but MO would have to spend an extra fs request (costing time) to find out if the file existed before.
And what if the file does exists in a mod? MO would have to copy the file to overwrite before it can open it for writing there. This would cause massive delays.
Worse: Many applications are lazy and open files with "full permissions" even when they're just reading so I might end up copying files to overwrite that are NOT being changed. Fun if that file is a 800MB bsa...
Worse yet: An existing file handle can be duplicated with increased permissions. Currently MO can ignore that but with the proposed change I would have to observe every handle duplication and see if  previously read-only handle is now writeable and then copy it. This means lots of coding, potential bugs and hit to performance.

And this is only the list of potential problems I can think of right now, experience says there is at least as many additional problems that I can't think of yet.

Also: You presume everyone uses MO like you do and wants new files in overwrite this is certainly not what everyone wants. I'd even argue this is a bad idea to use MO like this because overwrite is a shared folder between profiles (and I explained in the past why it has to be).
Having MO create new files in overwrite not only keeps your mods immutable, operations in one mod then affect mods that don't even include that mod.

Therefore this can't become the default. On the contrary: If this gets implemented it will be with a big fat warning sign: Enable only if you understand the consequences.

**Comment by grasmann** [comment]
I didn't know override is shared between profiles.
But then again I've never used different profiles.

IMO this is a bit of a philosophy question.
Personally I would consider the original structure of mods as holy, try to prevent any changes and shove everything that changes into override.
If that means very big files are copied, so be it.
I mean it's still possible to ask the user to replace a file directly if the size exceeds a certain limit.
If you have enough space on your hdds that shouldn't be a problem.

But of course I understand your concerns and that mod organizer is following another strategy.

**Comment by SilverIce** [comment]
yeah, the override folder is global, that's why I was proposing per-profile override folder in the past

anyway, I understand that any feature idea may tons of implementation issues, or not fit into the concept

**Comment by AnyOldName3** [comment]
As an opportunity for optimisation, what about recording differences to files in the cases writes are done directly to existing files without deleting them first? I'm aware there are a million standardised tools to do this for text files, and assume many would also work for binary files like esx files and bsa archives. Copying the original to the overwrite folder could be deferred until a later point (e.g. the program which was writing is terminated, or the user clicks a button to apply all write operations). Superficially, this doesn't look too complex (assuming Mod Organiser can actually see when a program writes to a file, which might not be the case), and dodges a lot of the performance issues, as copying is deferred as long as possible, and the mechanism only needs to kick in when a program actually writes to the VFS.

If file access is granted at a higher level, and MO can't see the data that passes through it, I could see this being a pain/impossible.

@SilverIce 
When I'm using BodySlide, I actually have a different MO profile which just has BodySlide and the mod whose outfits are being adjusted active. That way, there's nothing in the VFS for it to overwrite, and everything ends up in overwrite. I can move the resulting files to the correct place once I switch back to the default profile.

**Comment by patchling** [comment]
"When I'm using BodySlide, I actually have a different MO profile which just has BodySlide and the mod whose outfits are being adjusted active. That way, there's nothing in the VFS for it to overwrite, and everything ends up in overwrite. I can move the resulting files to the correct place once I switch back to the default profile."

i need to start doing this. i have my bodyslide generated file enabled/disabled when i am changing profiles, but this would be a no-accidental-overwrite method.

**Comment by TanninOne** [comment]
@AnyOldName3: I'm fairly certain recording binary changes at runtime would be incredibly, mindbogglingly slow. There are binary diff tools, sure. I just tested and doing a binary diff over 2 measly 10mb files takes a minute on my quad core i7. And this is not even considering that I would have to hook all file write/seek apis - which would hurt performance even more plus be further development effort.
Right now mo only affects file handle / search handle creation calls (that is: when file names are resolved), not the whole heavy duty apis and this is crucial to getting usable performance.

Regarding BodySlide: That is a neat idea, but it also only works if the override folder is empty. Otherwise you might overwrite files other profiles need there.

**Comment by AnyOldName3** [comment]
Keeping track of the changes might not be quite as slow as you fear. Were you to put in all that stuff that I now agree is probably far, far too much work, you wouldn't need to compare files at any point - the 'sensible' way to do things would be to record which address ranges in a file are written to, and what's written there. There's no need to perform any comparison with what was there already. I only mentioned existing diff tools as they'd provide existing file formats which could be used. Even when applying changes, we don't need to look at what was there already - we just overwrite it, making it not much slower than a copy, so as this can be done in a batch later on, it's an acceptable cost. When reads are made before changes are consolidated, we 'just' need to check if the address ranges being read from are listed in the diff file, and if so, serve up the modified data instead of the original, which again requires no comparison of the original. This last part is the only potentially slow bit that needs to happen at runtime.

**Comment by TanninOne** [comment]
@AnyOldName3: Sorry, but recording only the changes with the adress ranges would still be extremely slow as then I would be hooking and slowing down every file access (except for only the open&search operations). The hooking alone on functions like ReadFile will cause slowdowns I assume would be noticeable.
Also, when you say "we just need to check" you're overlooking that a read could include multiple change "chunks" interleaved with original data.
And this isn't even considering parallelism issues as the application might have multiple handles accessing the same file, potentially opened through different file names (hard links, junction points, ...). I would have to magically ensure all handles use the same modification "chunks" even if I don't know they refer to the same file. Plus what about thread-safety?
In short: The code would be very complex and error prone and slow.

I'm not going to implement this and my prediction is that even if someone else would, it would end up being unusable (buggy and/or slow).

Sorry, but this means we can't have fully immutable mods and I don't like to add an option "make mods a bit more immutable".
